tools/clang/unittests/HLSL/CompilerTest.cpp

```diff
@@ -2589,6 +2603,80 @@ TEST_F(CompilerTest, CompileWhenVdThenProducesDxilContainer) {
   VERIFY_IS_TRUE(hlsl::IsValidDxilContainer(reinterpret_cast<hlsl::DxilContainerHeader *>(pResultBlob->GetBufferPointer()), pResultBlob->GetBufferSize()));
 }
 
+void CompilerTest::TestEncodingImpl(const void *sourceData, size_t sourceSize, UINT32 codePage, const void *includedData, size_t includedSize, bool succ,
+                                    const WCHAR *encoding) {
+  CComPtr<IDxcCompiler> pCompiler;
+  CComPtr<IDxcOperationResult> pResult;
+  CComPtr<IDxcBlobEncoding> pSource;
+  CComPtr<TestIncludeHandler> pInclude;
+  VERIFY_SUCCEEDED(CreateCompiler(&pCompiler));
+  CreateBlobPinned((const char *)sourceData, sourceSize, codePage,
+                   &pSource);
+  pInclude = new TestIncludeHandler(m_dllSupport);
+  pInclude->CallResults.emplace_back(includedData, includedSize, 0);
+
+  const WCHAR *pArgs[] = {L"-encoding",
+                          encoding};
+  VERIFY_SUCCEEDED(pCompiler->Compile(pSource, L"source.hlsl", L"main",
+                                      L"ps_6_0", pArgs, _countof(pArgs),
+                                      nullptr, 0, pInclude, &pResult));
+  HRESULT status;
+  VERIFY_SUCCEEDED(pResult->GetStatus(&status));
+
+  if (succ) {
+    VERIFY_SUCCEEDED(status);
+  } else {
+    CComPtr<IDxcBlobEncoding> pErrors;
+    VERIFY_ARE_EQUAL(status, DXC_E_STRING_ENCODING_FAILED);
+    VERIFY_SUCCEEDED(pResult->GetErrorBuffer(&pErrors));
+    std::string textUtf8 = BlobToUtf8(pErrors);
+    LPCSTR pErrorMsg = "Error in encoding argument specified";
+    CheckOperationResultMsgs(pResult, &pErrorMsg, 1, false,
+                             false);
+  }
+}
+
+TEST_F(CompilerTest, CompileWithEncodeFlagTestSource) {
+  // Testing for utf8 encoded files
+  std::string sourceUtf8 = "#include \"include.hlsl\"\r\n"
+                             "float4 main() : SV_Target { return 0; }";
+  std::string includeUtf8 = "// Comment\n";
+  std::string utf8BOM = "\xEF" "\xBB" "\xBF"; // UTF-8 BOM
+  std::string includeUtf8BOM = utf8BOM + includeUtf8;
+
+  TestEncodingImpl(sourceUtf8.data(), sourceUtf8.size(), DXC_CP_UTF8,
+                   includeUtf8BOM.data(), includeUtf8BOM.size(),
+                   true, L"utf16");
+
+  TestEncodingImpl(sourceUtf8.data(), sourceUtf8.size(), DXC_CP_UTF8,
+                   includeUtf8.data(), includeUtf8.size(), false, L"utf16");
```

Comment:
So this is testing an invalid case where files are utf8 without BOM, and you say expect utf16.  Are you including it here because you get an encoding error?  Or is the size odd?  I think mismatch cases like this and the one below for the reverse can't be handled in any universal way really, so we should perhaps just skip the test?  Though if it currently fails slightly more gracefully due to your change to the error handling, I suppose this one could be ok.  I just feel like it shouldn't always be expected to detect mismatch for any arbitrary input.

Reply:
Previously there was a failed assert in checking the size, so this one of the locations where I changed it to throw a more specific error instead of a failed assert. That makes sense that it shouldn't be expected to detect mismatch for any arbitrary input, so I will remove it from the test cases.